### PR TITLE
llama : Add IBM granite template

### PR DESCRIPTION
**Pull request by arch-btw**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

Hi @ggerganov and @ngxson,

I'd like to contribute the IBM granite template to llama.cpp. 

I've made my best effort but I'm new to this, so could you please be so kind to review? Maybe you can see if it's alright. 

I've followed the wiki instructions but feel free to make changes or provide feedback on how to improve.

The model is here: https://huggingface.co/ibm-granite/granite-3.0-8b-instruct

Here is the current output of llama-cli:

> main: chat template example:
<|start_of_role|>system<|end_of_role|>
You are a helpful assistant<|end_of_text|>
<|start_of_role|>user<|end_of_role|>
Hello<|end_of_text|>
<|start_of_role|>assistant<|end_of_role|>
Hi there<|end_of_text|>
<|start_of_role|>user<|end_of_role|>
How are you?<|end_of_text|>
<|start_of_role|>assistant<|end_of_role|>

--

granite

### CODE REVIEW

File: src/llama.cpp [hunk 1]
```diff
@@ -21713,7 +21713,17 @@ static int32_t llama_chat_apply_template_internal(
                 ss << message->content << "\n\n";
             }
         }
-    } else {
+    } else if (tmpl == "granite" || tmpl_contains("<|start_of_role|>")) {
+    // IBM Granite template
+    for (const auto& message : chat) {
+        std::string role(message->role);
+        ss << "<|start_of_role|>" << role << "<|end_of_role|>" << "\n"
```
Comment (ngxson) [inline src/llama.cpp:21720]:
```suggestion
        ss << "<|start_of_role|>" << role << "<|end_of_role|>\n"
```

File: tests/test-chat-template.cpp [hunk 2]
```diff
@@ -65,6 +65,8 @@ int main(void) {
...
         // DeepSeek-V2
         "{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{{ bos_token }}{% for message in messages %}{% if message['role'] == 'user' %}{{ 'User: ' + message['content'] + '\n\n' }}{% elif message['role'] == 'assistant' %}{{ 'Assistant: ' + message['content'] + eos_token }}{% elif message['role'] == 'system' %}{{ message['content'] + '\n\n' }}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ 'Assistant:' }}{% endif %}",
+        // ibm-granite/granite-3.0-8b-instruct
+        "<|start_of_role|>system<|end_of_role|>You are a helpful assistant<|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Hello<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>Hi there<|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Who are you<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>   I am an assistant   <|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Another question<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>",
```
Comment (ngxson) [inline tests/test-chat-template.cpp:69]:
```suggestion
        "<|start_of_role|>system<|end_of_role|>You are a helpful assistant<|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Hello<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>Hi there<|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Who are you<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>   I am an assistant   <|end_of_text|>\n<|start_of_role|>user<|end_of_role|>Another question<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>\n",
```

Probably missing a new line at the end - please double check on your side

**Review by ngxson — CHANGES_REQUESTED** [review]

File: tests/test-chat-template.cpp (hunk 2 above)
Comment (ngxson) [reply tests/test-chat-template.cpp:69]: You can also run chat template test locally:

```
make tests/test-chat-template && ./tests/test-chat-template
```

**Comment by arch-btw** [comment]
@ngxson thank you so much for your help, I appreciate your time and effort!

I have applied your review 👍 

I have a quick question, when I run ```test-chat-template``` is it normal that it looks like this (screenshot below).
The part that confuses me is that <end_of_sentence> is not escaped but the other tags are.
And then user/assistant are on the same line, twice.

I'm not sure what it should look like in the context of this script, considering the other prompts have things like additional spacing as well.

test-chat-template

**Comment by ngxson** [comment]
You need to add an example up unformatted chat template in `test-chat-template`. Please look carefully in the file.

**Comment by arch-btw** [comment]
Hi @ngxson 

Thank you again for your guidance!
I finally managed to get it to work! 😅 

Here is the output of ``make tests/test-chat-template && ./tests/test-chat-template``:

-----

outputgranite

-----

And here is the output of ``./llama-cli -m granite-3.0-8B-instruct-F32-Q5_K_M.gguf -t 4 -p 'You are granite, an AI model.' --conversation --color --chat-template granite``:

outputgranite2

And a prompt example:

outputgranite3

File: src/llama.cpp [hunk 3]
```diff
@@ -21713,7 +21713,17 @@ static int32_t llama_chat_apply_template_internal(
...
+    // IBM Granite template
+    for (const auto& message : chat) {
+        std::string role(message->role);
+        ss << "<|start_of_role|>" << role << "<|end_of_role|>"
+           << message->content << "<|end_of_text|>\n";
+    }
+    if (add_ass) {
+        ss << "<|start_of_role|>assistant<|end_of_role|>\n";
+    }
+} else {
```
Comment (ngxson) [inline src/llama.cpp:21726]: small nits before I can merge: the indentation here is incorrect (missing one level)

File: src/llama.cpp [hunk 4]
```diff
@@ -21713,7 +21713,17 @@ static int32_t llama_chat_apply_template_internal(
                 ss << message->content << "\n\n";
             }
         }
-    } else {
+    } else if (tmpl == "granite" || tmpl_contains("<|start_of_role|>")) {
+    // IBM Granite template
+    for (const auto& message : chat) {
```
Comment (ngxson) [inline src/llama.cpp:21718]: code style:

```suggestion
    for (const auto & message : chat) {
```

**Review by ngxson — COMMENTED** [review]
Small changes before I can merge

**Comment by arch-btw** [comment]
@ngxson Thank you again and sorry about that!

File: src/llama.cpp [hunk 5]
```diff
@@ -21713,6 +21713,16 @@ static int32_t llama_chat_apply_template_internal(
                 ss << message->content << "\n\n";
             }
         }
+    } else if (tmpl == "granite" || tmpl_contains("<|start_of_role|>")) {
+        // IBM Granite template
+        for (const auto & message: chat) {
+            std::string role(message -> role);
```
Comment (ngxson) [inline src/llama.cpp:21719]:
```suggestion
            std::string role(message->role);
```

File: src/llama.cpp [hunk 6]
```diff
@@ -21713,6 +21713,16 @@ static int32_t llama_chat_apply_template_internal(
...
+            ss << "<|start_of_role|>" << role << "<|end_of_role|>"
+               << message -> content << "<|end_of_text|>\n";
```
Comment (ngxson) [inline src/llama.cpp:21721]:
```suggestion
               << message->content << "<|end_of_text|>\n";
```

**Review by ngxson — APPROVED** [review]
Thanks, merging once the CI passes

File: src/llama.cpp [hunk 7]
```diff
@@ -21713,6 +21713,16 @@ static int32_t llama_chat_apply_template_internal(
                 ss << message->content << "\n\n";
             }
         }
+    } else if (tmpl == "granite" || tmpl_contains("<|start_of_role|>")) {
+        // IBM Granite template
+        for (const auto & message: chat) {
```
Comment (ngxson) [inline src/llama.cpp:21718]:
```suggestion
        for (const auto & message : chat) {
```

**Pull request merged**